### ISSUE TITLE
Help with the documentation

### ISSUE THREAD

**Issue by claudio20497** [JuliaRobotics/KernelDensityEstimate.jl#67]
Hello,

I am interested in using your package, but I am not a domain expert in kde estimation or products of them.
From the ReadMe it is not clear to me what methods I may call on a BallTreeDensity. For example, I noted that calling `rand` on a BallTreeDensity like this:

```julia
extractions = randn(1000)
p = kde!(extractions )
rand(p)
```

actually works.

1. **Could you add to the ReadMe a list of the methods one may call on a BallTreeDensity?**
2. **More specifically, what does the `resample` method do?**
3. **Is there a way to evaluate the pdf of a BallTreeDensity at a point, even when this point is not included in the dataset from which we fit the kde?** I mean something KernelDensity.jl - like ( using the `p` from before):

```julia
pdf(p, 0.5) # evaluate the probability density of p at 0.5 ( even though 0.5 was not included in `extractions`)
```

Regarding question 3. , I saw [this](https://github.com/JuliaRobotics/KernelDensityEstimate.jl/issues), but I didn't really understand. 

Great package!

Thanks in advance

**Comment by dehann** [comment]
Hi @claudio20497 ,

Thanks for posting and suggestions.  I will add as soon I can, but in the mean time:
- List of calls on `BallTreeDensity` should not be too long, mostly check list of export functions that take `::BallTreeDensity`: https://github.com/JuliaRobotics/KernelDensityEstimate.jl/blob/master/src/KernelDensityEstimate.jl
- `resample` takes any kde of M samples and then builds a new kde consisting of N sample number requested by user,
- kernel bandwidths are "diagonals only", but there is an independent variance along each dimension.  No covariance (off-diagonals).  This is an improvement over original library that assumed a single isotropic bandwidth across all dimensions.  Using only diagonals are not so bad since the idea is approximate the pdf function -- i.e. adding together many 'hyper-ellipses' can still approximate an overall belief function of any orientation and weight distribution arbitrarily.  Relaxing each kernel to a complete dense covariance matrix might become tricky (if i recall the internals correctly).  I'd suggest experimenting with a simple minimum example and try 'break' the approximation in a way where more samples in the kde are not good enough and in a way where having off-diagonal kernel bandwidths would have helped.  There might be performance arguments this way or that way.  This package currently is more of a kde and product of kdes package, rather than drop in replacement for low mixture count Gaussian mixture model (GMM) -- i.e. if you have 5 modes in a belief, then use 100+ samples.
- Yes, it should be easy to evaluate the pdf anywhere:
```julia
# build a random kde object
X = kde!(randn(2,100))

# eval on the objection itself to get pdf density values
densities = X(5*rand(2,10))

# see plotting at KernelDensityEstimatePlotting.jl for examples
```

**Comment by claudio20497** [comment]
Hello @dehann ,

Thank you very much for the detailed answer.

So concerning point 3,  if I understood correctly, if I sample from - say - a bivariate distribution of two correlated variables, and then call `kde!`  on that sample, the resulting `BallTreeDensity` won't exhibit the correlation again, right?

**Comment by dehann** [comment]
> Yes, correlations are conserved.

That's correct, the correlations are conserved.  This remains true even though the individual kernel bandwidths that make up the kde use diagonal only values.

Also note KernelDesityEstimatePlotting package exists with useful function `plotKDE` and a variety of keyword options.

**Comment by ExpandingMan** [comment]
I just wanted to step in and mention that I discovered this package today, it looks quite nice, but I think lack of documentation is going to make it quite difficult for me to use.  Even a link to a review of the algorithms involved would be enormously helpful, coming in cold it's very unclear what most of these methods are doing.
